Deprecate /containers/(id or name)/copy endpoint

vdemeester opened the pull request:
This endpoint has been deprecated since 1.8. Return an error starting from this API version (1.24) in order to make sure it's not used for the next API version and so that we can remove it some times later  🐮.

This is probably the cleanest and simple way to handle that.

🐸

Signed-off-by: Vincent Demeester vincent@sbr.pm

vdemeester commented on api/server/router/container/copy.go:
```diff
@@ -16,6 +16,10 @@ import (
 func (s *containerRouter) postContainersCopy(ctx context.Context, w http.ResponseWriter, r *http.Request, vars map[string]string) error {
+	version := httputils.VersionFromContext(ctx)
+	if version.GreaterThanOrEqualTo("1.24") {
+		return fmt.Errorf("/containers/(id or name)/copy is not supported for this API version : %s", version)
```
Not sure this is the right error — as discussed a little bit with @thaJeztah, a simple HTTP 404 error should be fine.

thaJeztah commented:
In general, SGTM, we can discuss the implementation, and we need to have a note in the API changelog

vdemeester commented on api/server/router/container/copy.go:
```diff
@@ -16,6 +16,11 @@ import (
 // postContainersCopy is deprecated in favor of getContainersArchive.
 func (s *containerRouter) postContainersCopy(ctx context.Context, w http.ResponseWriter, r *http.Request, vars map[string]string) error {
+	version := httputils.VersionFromContext(ctx)
+	if version.GreaterThanOrEqualTo("1.24") {
```
I probably should add a comment here (instead of in `api/server/router/container/container.go`

thaJeztah replied:
👍

thaJeztah commented:
moved this to code review; can you update the comment and documentation?

vdemeester commented:
Updated :angel:

icecrime commented on integration-cli/docker_api_containers_test.go:
```diff
@@ -1032,7 +1032,22 @@ func (s *DockerSuite) TestContainerApiCopy(c *check.C) {
 	}
 
-	status, body, err := sockRequest("POST", "/containers/"+name+"/copy", postData)
+	status, _, err := sockRequest("POST", "/vcontainers/"+name+"/copy", postData)
```
`/vcontainers/`?

vdemeester replied:
damn me :sob:

thaJeztah commented:
LGTM, don't forget to remove it from the API docs; https://github.com/docker/docker/blob/master/docs/reference/api/docker_remote_api_v1.24.md#copy-files-or-folders-from-a-container

vdemeester commented:
@thaJeztah ahh good point :blush: Updated :angel:

thaJeztah commented:
LGTM (code and docs)

thaJeztah commented:
ping @vdemeester sorry, this needs a rebase :cry:

thaJeztah commented:
oh! actually, can you add an entry to the "deprecated features" page? This endpoint was deprecated before we had that document, so I think we should add a mention there.

Also just discussing with @justincormack that we should have a deprecation policy for old API version (how many versions should we support?) otherwise removing these endpoints won't bring us much, because we still have to maintain it if we keep the old API versions

vdemeester commented:
@thaJeztah ok 😝 and completely agree with the depreciation policy 😇

vdemeester committed:
Deprecate /containers/(id or name)/copy endpoint

This endpoint has been deprecated since 1.8. Return an error starting
from this API version (1.24) in order to make sure it's not used for the
next API version and so that we can remove it some times later.

Signed-off-by: Vincent Demeester <vincent@sbr.pm>

thaJeztah commented:
ping @justincormack @calavera PTAL

runcom commented:
LGTM

vdemeester commented:
Author docs LGTM :stuck_out_tongue_closed_eyes: 
Oh I can do that right ? :angel:
